src/catchup/CatchupManagerImpl.cpp

```diff
@@ -28,32 +28,143 @@ CatchupManager::create(Application& app)
 }
 
 CatchupManagerImpl::CatchupManagerImpl(Application& app)
-    : mApp(app), mCatchupWork(nullptr)
+    : mApp(app)
+    , mCatchupWork(nullptr)
+    , mSyncingLedgersSize(
+          app.getMetrics().NewCounter({"ledger", "memory", "queued-ledgers"}))
 {
 }
 
 CatchupManagerImpl::~CatchupManagerImpl()
 {
 }
 
+uint32_t
+getCatchupCount(Application& app)
+{
+    return app.getConfig().CATCHUP_COMPLETE
+               ? std::numeric_limits<uint32_t>::max()
+               : app.getConfig().CATCHUP_RECENT;
+}
+
 void
-CatchupManagerImpl::historyCaughtup()
+CatchupManagerImpl::processLedger(LedgerCloseData const& ledgerData)
 {
-    mCatchupWork.reset();
+    uint32_t lastReceivedLedgerSeq = ledgerData.getLedgerSeq();
+
+    auto addLedgerData = [&]() { addToSyncingLedgers(ledgerData); };
+
+    // mSyncingLedgers can be non-empty after mCatchupWork finishes if there was
+    // a failure during catchup
+    if (mCatchupWork && mCatchupWork->isDone())
+    {
+        if (!mSyncingLedgers.empty())
+        {
+            addLedgerData();
```

Comment:
Can we maybe write the code as to only have 3 steps to simplify things?
I don't like that we special case catchup retries.

1. manage `mCatchupWork` (ie, "reset" if done)
2. update `mSyncingLedgers` as needed (append, trim if `mCatchupWork` is not running)
3. start `mCatchupWork` if needed

Reply:
created `trimSyncingLedgers` function and simplified retry logic.